### PR TITLE
Copying the audio engine implementation for Windows to Linux

### PR DESCRIPTION
**Pull request by JOOpdenhoevel**
The title says what this is about: I have copied the audio engine for Windows to Linux and made it work. I did this because the Windows implementation uses OpenAL instead of FMOD and
it was quite strange to me that Linux uses a proprietary audio library, while every other platform uses the open one. Also, I think that the audio engine is easier to maintain when every platform uses the same API.

I was surprised that this was such an easy job because I only needed to copy the source files, change every "Win32" to "Linux" and tell CMake to link against these libraries. I wonder why no one did this before. Anyway, why was the audio engine implemented with FMOD and not with OpenAL?

I also updated the installation script, but the deps-archive still contains FMOD. It would be really nice if someone could remove it when this get merged.

EDIT: The test failed since the dependencies for linux have changed.

### CODE REVIEW

**Comment by minggo** [comment]
[ci rebuild]

**Comment by ricardoquesada** [comment]
[ci rebuild]

**Comment by ricardoquesada** [comment]
I'll take a look at this one.

**Comment by ricardoquesada** [comment]
@minggo 
the PR is failing because I can't find some libraries.
what script should I edit to tell the builder to install the missing libraries?

in particular:

```
-- Vorbis include dirs: 
CMake Error at cmake/Modules/FindPackageHandleStandardArgs.cmake:136 (message):
  Could NOT find MPG123 (missing: MPG123_INCLUDE_DIR MPG123_LIBRARY)
Call Stack (most recent call first):
```

**Comment by minggo** [comment]
@ricardoquesada i think the codes `cocos_find_package(MPG123 MPG123 REQUIRED)` will find it. Did you invoke `build/install-linux-deps.sh` to install MPG?

**Comment by ricardoquesada** [comment]
@minggo I mean, I think the builder is failing because the builder doesn't have installed those packages. So, I have to tell the builder to install the packages.

**Comment by minggo** [comment]
@zilongshanren will the builder run the `install-linux-deps` for every PR?

**Comment by minggo** [comment]
@ricardoquesada i asked @zilongshanren help to manually installed it.

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by ricardoquesada** [comment]
@minggo thanks!

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by ricardoquesada** [comment]
I've just tested this PR on my Linux machine and everything seems to be working ok except that in the "7 - Audio CocosDenshion", it fails with the "Control effects" with this error:

`Unsupported media type file: effect1.wav`

**Comment by ricardoquesada** [comment]
[ci rebuild]

**Comment by JOOpdenhoevel** [comment]
@ricardoquesada Themen Cocos Denshion Test uses the simple Audio engine which uses a standart Windows library which propably supports .wav. The simple a.e. for Linux uses the proper audio engine, which is why it doesn't support that. I'm going to fix that.

**Comment by ricardoquesada** [comment]
@Janonard great. thanks!

**Comment by JOOpdenhoevel** [comment]
I guess I need some help: I would like to use FFmpeg's avcodec, since it supports almost everything. How do I tell cmake to link against it? I'm quite new to the project and I'm not yet used to cocos' cmake. May I get some help?

**Comment by JOOpdenhoevel** [comment]
As mentioned above, I need to add another library to support .wav files. When I think about it, it makes more sense to get the merging done and add the library afterward. It appears to me that adding a new dependency is quite a big thing for Cocos and I assume that it is better to discuss which library to use and which files to support later. I will do that, but I guess it's better to fulfill the cause of this pull request and add other file formats then. Is that okay for you?

**Comment by JOOpdenhoevel** [comment]
I merged the audio engine for Windows and for Linux into the "Desktop" implementation. The tests for the main audio engine are positive, but the ones for the simple audio engine are broken because the S.A.E. is a wrapper for the new A.E. and it doesn't support the file formats  used in the test. It was like this on Linux all the time and the new engine does everything the old does while beeing easier to extend. Because of this I suggest do abandon the old engine. Do you subscribe my point of view?

**Comment by ricardoquesada** [comment]
@Janonard 
Yes, SimpleAudioEngine will be deprecated in v4.0. But we can't deprecate it right now, not in a Point release.
Regarding the wav file, we should add support for it before merging this PR since it was working before. Adding "FFmpeg" dependency is not a problem in linux, assuming that it is a package that users can install using "apt-get" or similar.

**Comment by ricardoquesada** [comment]
@Janonard should we also include Mac in the "desktop" thing? I mean, use this same for Mac? (what are we using on Mac BTW?)

**Comment by JOOpdenhoevel** [comment]
@ricardoquesada

> Yes, SimpleAudioEngine will be deprecated in v4.0.

Okay, good to know.

> Regarding the wav file, we should add support for it.

OK, will do that, but this will take some more time since I need to familiarize myself in the CMake setup and in FFmpeg.

> Should we also include Mac in the "desktop" thing?

From my point of view, Mac should be included, but I can't do this since I don't own a Mac or have the money to buy one. Someone else would need to do that.

**Comment by ricardoquesada** [comment]
@Janonard 

> will do that, but this will take some more time since I need to familiarize myself in the 
> CMake setup and in FFmpeg.

if you just make it work (the fixes in the code), I can fix the cmake stuff and needed dependencies.

> From my point of view, Mac should be included, but I can't do this since I don't own a Mac or have the money to buy one. Someone else would need to do that.

Ok. I can fix that I guess.

**Comment by JOOpdenhoevel** [comment]
@ricardoquesada 
I'm having a problem: I trying to test FFmpeg in a new CMake project, but linking to FFmpeg simply doesn't work, the linker complains that some symbols haven't been resolved, although I linked against FFmpeg. I tried all possible options for CMake, but it even doesn't work from the command line. Can you figure that out, please? It would be nice if you would add FFmpeg to cocos and send me a pull request with it. :)

**Comment by ricardoquesada** [comment]
@Janonard sure. I can figure out how to link it. Have you added the ffmpeg code? It doesn't matter if doesn't link, but in order to fix the linking issue I need to try with code that uses it.

**Comment by JOOpdenhoevel** [comment]
@ricardoquesada 
The problem is because I couldn't link against FFmpeg in any way, I couldn't experiment with and therefore I couldn't learn how to use it. I have added a small and silly test that uses libavformat, libavcodec and libavutil, which are the libraries I'm probably going to use, which prints the gcd of 4 and 8, the version of avformat and the version of avcodec into the log. I'm going to remove it when you're done, but you can do that too if you like.

**Comment by ricardoquesada** [comment]
@Janonard ok. got it. I'll add ffmpeg first.

File: cocos/audio/CMakeLists.txt [hunk 1]
```diff
@@ -3,14 +3,12 @@ set(COCOS_AUDIO_SRC
     audio/AudioEngine.cpp
     )
 
-if(WINDOWS)
+if(LINUX OR WINDOWS)
     set(COCOS_AUDIO_PLATFORM_SRC
-        audio/win32/SimpleAudioEngine.cpp
-        audio/win32/MciPlayer.cpp
-        audio/win32/MciPlayer.h
-        audio/win32/AudioEngine-win32.cpp
```
Comment (ricardoquesada) [inline cocos/audio/CMakeLists.txt:11]: it seems that `MciPlayer.cpp` was removed from windows. was this done on purpose? or is it just a copy&paste error?

**Comment by JOOpdenhoevel** [comment]
Yes, this is intended. MciPlayer used a standart Windows library and therefore I couldn't keep it. The simple audio engine for Windows was a standalone implementation while the SAE for Linux wrapped the new one. I decided to go with the Linux way.

**Comment by minggo** [comment]
@ricardoquesada any update about this PR? Can i merge it when it passes CI and fix conflict?

**Comment by ricardoquesada** [comment]
@minggo there is one bug that needs to be fixed. but in order to do that I told the @Janonard that I was going to add CMake support for FFMpeg, but I didn't have the time to do it yet.

**Comment by JOOpdenhoevel** [comment]
I am very sorry, but I am not interested in this topic anymore. While I have waited for the FFmpeg integration, I began to work on a new project, which is near public release right now. Since I have to focus on this, I have no more time for cocos2d-x, maybe I will return to it later.

**Comment by v1993** [comment]
@drelaptop @minggo Can you finish and merge this pull request, please?

**Comment by drelaptop** [comment]
@v1993  it's a same issue with https://github.com/cocos2d/cocos2d-x/issues/16355 https://github.com/cocos2d/cocos2d-x/issues/15419, I think we should finish it at 3.18 milestone, what do you think? @minggo

**Comment by minggo** [comment]
Yep, i think we can do it in 3.18 because FMOD always cause issues on linux.

**Pull request closed without merging**